include/net/net_timeout.h

```diff
@@ -23,38 +24,244 @@
 #endif
 
-/** Let the max timeout be 100 ms lower because of
- * possible rounding in delayed work implementation.
+/** @brief Divisor used to support ms resolution timeouts.
+ *
+ * Because delays are processed in work queues which are not invoked
+ * synchronously with clock changes we need to be able to detect timeouts
+ * after they occur, which requires comparing "deadline" to "now" with enough
+ * "slop" to handle any observable latency due to "now" advancing past
+ * "deadline".
+ *
+ * The simplest solution is to use the native conversion of the well-defined
+ * 32-bit unsigned difference to a 32-bit signed difference, which caps the
+ * maximum delay at INT32_MAX.  This is compatible with the standard mechanism
+ * for detecting completion of deadlines that do not overflow their
+ * representation.
  */
-#define NET_TIMEOUT_MAX_VALUE ((uint32_t)(INT32_MAX - 100))
+#define NET_TIMEOUT_MAX_VALUE ((uint32_t)INT32_MAX)
 
-/** Generic struct for handling network timeouts */
+/** Generic struct for handling network timeouts.
+ *
+ * Except for the linking node, all access to state from these objects must go
+ * through the defined API.
+ */
 struct net_timeout {
-	/** Used to track timers */
+	/** Used to link multiple timeouts that share a common timer infrastructure.
+	 *
+	 * For examples a set of related timers may use a single delayed work
+	 * structure, which is always scheduled at the shortest time to a
+	 * timeout event.
+	 */
 	sys_snode_t node;
 
-	/** Address lifetime timer start time */
+	/* Time at which the timer was last set.
+	 *
+	 * This usually corresponds to the low 32 bits of k_uptime_get(). */
 	uint32_t timer_start;
 
-	/** Address lifetime timer timeout in milliseconds. Note that this
-	 * value is signed as k_delayed_work_submit() only supports signed
-	 * delay value.
+	/* Portion of remaining timeout that does not exceed
+	 * NET_TIMEOUT_MAX_VALUE.
+	 *
+	 * This value is updated in parallel with timer_start and wrap_counter
+	 * by net_timeout_evaluate().
 	 */
-	int32_t timer_timeout;
+	uint32_t timer_timeout;
 
-	/** Timer wrap count. Used if the timer timeout is larger than
-	 * about 24 days. The reason we need to track wrap arounds, is
-	 * that the timer timeout used in k_delayed_work_submit() is
-	 * 32-bit signed value and the resolution is 1ms.
+	/* Timer wrap count.
+	 *
+	 * This tracks multiples of NET_TIMEOUT_MAX_VALUE milliseconds that
+	 * have yet to pass.  It is also updated along with timer_start and
+	 * wrap_counter by net_timeout_evaluate().
 	 */
-	int32_t wrap_counter;
+	uint32_t wrap_counter;
 };
 
+/** @brief Configure a network timeout structure.
+ *
+ * @param timeout a pointer to the timeout state.
+ *
+ * @param lifetime the duration of the timeout in seconds.
+ *
+ * @param now the time at which the timeout started counting down, in
+ * milliseconds.  This is generally a captured value of k_uptime_get_32().
+ */
+static inline void net_timeout_set(struct net_timeout *timeout,
```

Comment:
Why inline for these new functions?

Reply:
Mostly because I didn't want to introduce a new source file to hold them.  Are you OK with having the implementation for this header in the `net_if.c` file?

Reply:
Ah, indeed. Well, perhaps we could introduce `net_timeout.c` in this case.

Reply:
Hmn.  OK.  I doubt anything will use without having `net_if.c` as well; the other in-tree use is in ipv6.

I can add the implementation file, it does make the unit test a little more invasive though as it has to include the implementation as well as the header.

Reply:
We can certainly put these to net_if.c if it looks too complicated, no worries.

Reply:
No, if they're not inline it's best they be in a separate file.  `net_if.c` wouldn't work as a unit test, which doesn't use the standard Zephyr build infrastructure.